### PR TITLE
fix!: add rust tokenizer support with new parser

### PR DESCRIPTION
**Pull request by tobymao**

### CODE REVIEW

File: sqlmesh/core/dialect.py [hunk 1]
```diff
@@ -98,32 +98,6 @@ class StagedFilePath(exp.Table):
     """Represents paths to "staged files" in Snowflake."""
 
 
-def _scan_var(self: Tokenizer) -> None:
```
Comment (georgesittas) [inline sqlmesh/core/dialect.py:101]: so we basically got rid of this and now it's handled at parse time?

Comment (tobymao) [reply sqlmesh/core/dialect.py:101]: yea

File: sqlmesh/core/dialect.py [hunk 2]
```diff
@@ -150,6 +124,52 @@ def _parse_lambda(self: Parser, alias: bool = False) -> t.Optional[exp.Expressio
...
+    self: Parser,
+    any_token: bool = True,
+    tokens: t.Optional[t.Collection[TokenType]] = None,
+) -> t.Optional[exp.Expression]:
+
+    if self._match(TokenType.L_BRACE):
+        identifier = self.__parse_id_var(any_token=any_token, tokens=tokens)  # type: ignore
+        if not self._match(TokenType.R_BRACE):
+            self.raise_error("Expecting }")
+        identifier.args["this"] = f"@{{{identifier.args['this']}}}"
```
Comment (georgesittas) [inline sqlmesh/core/dialect.py:137]:
```suggestion
        identifier.args["this"] = f"@{{{identifier.name}}}"
```

File: sqlmesh/core/dialect.py [hunk 3]
```diff
@@ -150,6 +124,52 @@ def _parse_lambda(self: Parser, alias: bool = False) -> t.Optional[exp.Expressio
...
+
+    while (
+        identifier
+        and self._is_connected()
+        and (
+            self._match_texts(("{", SQLMESH_MACRO_PREFIX), advance=False)
+            or self._curr.token_type not in self.RESERVED_TOKENS
+        )
+    ):
+        this = identifier.args["this"]
```
Comment (georgesittas) [inline sqlmesh/core/dialect.py:149]:
```suggestion
        this = identifier.name
```

File: sqlmesh/core/dialect.py [hunk 4]
```diff
@@ -150,6 +124,52 @@ def _parse_lambda(self: Parser, alias: bool = False) -> t.Optional[exp.Expressio
...
+        brace = self._match(TokenType.L_BRACE)
+
+        if brace:
+            this += "{"
+
+        next_id = self._parse_id_var(any_token=False)
+
+        if next_id:
+            this += next_id.this
```
Comment (georgesittas) [inline sqlmesh/core/dialect.py:158]: here too I guess?
```suggestion
            this += next_id.name
```

File: sqlmesh/core/dialect.py [hunk 5]
```diff
@@ -150,6 +124,52 @@ def _parse_lambda(self: Parser, alias: bool = False) -> t.Optional[exp.Expressio
...
+        else:
+            return identifier
+
+        if brace:
+            if self._match(TokenType.R_BRACE):
+                this += "}"
+            else:
+                self.raise_error("Expecting }")
+
```
Comment (georgesittas) [inline sqlmesh/core/dialect.py:167]: Note: maybe out of scope for now, but I believe we have this pattern of "assert we match something" in sqlglot's parser as well -- perhaps we could factor it out using an extra flag in `_match`?

```python
self._match(TokenType.R_BRACE, ensure_matched=True)
```

**Review by georgesittas — APPROVED** [review]

**Pull request merged**